internal/privatestate/data.go

```diff
@@ -0,0 +1,363 @@
+
+				continue
+			}
+
+			mergedMap[k] = v
+		}
+	}
+
+	if diags.HasError() {
+		return nil, diags
+	}
+
+	bytes, err := json.Marshal(mergedMap)
+	if err != nil {
+		diags.AddError(
+			"Error Encoding Private State",
+			fmt.Sprintf("An error was encountered when encoding private state: %s.\n\n"+
+				"This is always a problem with Terraform or terraform-plugin-framework. Please report this to the provider developer.", err),
+		)
+
+		return nil, diags
+	}
+
+	return bytes, diags
+}
+
+// NewData creates a new Data based on the given slice of bytes.
+// It must be a JSON encoded slice of bytes, that is map[string][]byte.
+func NewData(ctx context.Context, data []byte) (*Data, diag.Diagnostics) {
+	var (
+		dataMap map[string][]byte
+		diags   diag.Diagnostics
+	)
+
+	if len(data) == 0 {
+		return nil, nil
+	}
+
+	err := json.Unmarshal(data, &dataMap)
+	if err != nil {
+		diags.AddError(
+			"Error Decoding Private State",
+			fmt.Sprintf("An error was encountered when decoding private state: %s.\n\n"+
+				"This is always a problem with Terraform or terraform-plugin-framework. Please report this to the provider developer.", err),
+		)
+
+		return nil, diags
+	}
+
+	output := Data{
+		Framework: make(map[string][]byte),
+		Provider: &ProviderData{
+			make(map[string][]byte),
+		},
+	}
+
+	for k, v := range dataMap {
+		if !utf8.Valid(v) {
+			diags.AddError(
+				"Error Decoding Private State",
+				"An error was encountered when validating private state value.\n"+
+					fmt.Sprintf("The value being supplied for key %q is is not valid UTF-8.\n\n", k)+
+					"This is always a problem with Terraform or terraform-plugin-framework. Please report this to the provider developer.",
+			)
+
+			tflog.Error(ctx, "error decoding private state: invalid UTF-8 value", map[string]interface{}{"key": k, "value": v})
+
+			continue
+		}
+
+		if !json.Valid(v) {
+			diags.AddError(
+				"Error Decoding Private State",
+				"An error was encountered when validating private state value.\n"+
+					fmt.Sprintf("The value being supplied for key %q is is not valid JSON.\n\n", k)+
+					"This is always a problem with Terraform or terraform-plugin-framework. Please report this to the provider developer.",
+			)
+
+			tflog.Error(ctx, "error decoding private state: invalid JSON value", map[string]interface{}{"key": k, "value": v})
+
+			continue
+		}
+
+		if isInvalidProviderDataKey(ctx, k) {
+			output.Framework[k] = v
+			continue
+		}
+
+		output.Provider.data[k] = v
+	}
+
+	if diags.HasError() {
+		return nil, diags
+	}
+
+	return &output, diags
+}
+
+// NewProviderData creates a new ProviderData based on the given slice of bytes.
+// It must be a JSON encoded slice of bytes, that is map[string][]byte.
+func NewProviderData(ctx context.Context, data []byte) (*ProviderData, diag.Diagnostics) {
+	providerData := &ProviderData{
+		data: make(map[string][]byte),
+	}
```

Comment:
World's smallest nit in case the logic is ever more complicated 😎 

```suggestion
	providerData := EmptyData(ctx)
```